📖 Update the Operator documentation and FAQ (#47)
hashicorp/terraform-cloud-operator: docs/faq.md

- **What will happen if I have multiple deployments of the Operator watching the same namespace(s)?**

  Behavior may vary from controller to controller. Most probably you will notice that Customer Resource objects are constantly reconciled and that might cause constant updates of Terraform Cloud objects. For example, the `Module` controller might trigger a new run every reconciliation and because of that the Run queue will grow infinitely and consume all resources.

  It is definitely better to avoid such situations.

- **What do the `*-workers` options?**

  The `*-workers` options configure the number of specific controller workers that are concurrently processing Customer Resources. In certain cases increasing this number can improve performance.

- **What does the `sync-period` option?**

>   The sync-period option specifies the minimum frequency at which watched resources are reconciled. That is necessary to make sure that managed resource is in the desired state. The synchronization period should be aligned with the number of managed Customer Resources. If the period is too low and the number of managed resources is too high, you may observe slowness in synchronization.

Reviewer: ```suggestion
  The `--sync-period` option specifies the minimum frequency at which watched resources are reconciled. The synchronization period should be aligned with the number of managed Customer Resources. If the period is too low and the number of managed resources is too high, you may observe slowness in synchronization.
```

## Performance

- **How many Custom Resources can be managed by a single deployment of the Operator?**

  In theory, a single deployment of the Operator can manage thousands of Customer Resources, but more accurately it would be to answer how many Customer Resources can be managed with a single token. The Operator's performance depends on the number of API calls it does and the Terraform Cloud API [rate limit](https://developer.hashicorp.com/terraform/cloud-docs/api-docs#rate-limiting).

  The number of API calls the Operator does depends on multiple factors:

    - the value of the `sync-period` option;

    - the values of `*-workers` options;

>     - type of the Customer Resources;

Reviewer: ```suggestion
    - The type of the resource.
```

>     - the features managed by Customer Resources.

Reviewer: ```suggestion
    - The Terraform Cloud features being used.
```

>   With the default values of `sync-period`(5 minutes) and `*-workers`(1 worker per controller), we recommend managing 100 Customer Resources per token. This number can vary based on previously mentioned factors. The number can be updated later with new versions or changes in the Terraform Cloud API.

Reviewer: ```suggestion
  With the default values of `sync-period` (5 minutes) and `*-workers` (1 worker per controller), we recommend managing **100 resources per token**. This number can vary based on previously mentioned factors. This number can be updated later to accommodate  changes in the Terraform Cloud API.
```

## Workspace Controller

- **Can a single deployment of the Operator manage the Workspaces of different Organizations?**

>   Yes, it can. Workspace Customer Resource has mandatory fields `spec.organization` and `spec.token`. The Operator manages workspaces based on these credentials.

Reviewer: ```suggestion
  Yes. The Workspace resource has mandatory fields `spec.organization` and `spec.token`. The Operator manages workspaces based on these credentials.
```

- **Where can I find Workspace outputs?**

>   Non-sensitive outputs will be saved in Kubernetes ConfigMap. Sensitive outputs will be saved in Kubernetes Secret. In both cases, the name of the corresponding Kubernetes object will be generated automatically and has the following pattern: `<CR.metadata.name>-outputs`.

Reviewer: ```suggestion
  Non-sensitive outputs will be saved in a ConfigMap. Sensitive outputs will be saved in a Secret. In both cases, the name of the corresponding Kubernetes resource will be generated automatically and has the following pattern: `<metadata.name>-outputs`.
```

## Module Controller

- **Where can I find Module outputs?**

>   Non-sensitive outputs will be saved in Kubernetes ConfigMap. Sensitive outputs will be saved in Kubernetes Secret. In both cases, the name of the corresponding Kubernetes object will be generated automatically and has the following pattern: `<CR.metadata.name>-module-outputs`. In the case the underlying Workspace is managed by the Operator, all outputs will be duplicated in the corresponding ConfigMap/Secret.

Reviewer: ```suggestion
  Non-sensitive outputs will be saved in a ConfigMap. Sensitive outputs will be saved in a Secret. In both cases, the name of the corresponding Kubernetes object will be generated automatically and has the following pattern: `<metadata.name>-module-outputs`. When the underlying workspace is managed by the operator, all outputs will be duplicated in the corresponding ConfigMap or Secret.
```

- **Can I execute a new Run without changing any Workspace or Module attributes?**

>   Yes, you can. There is a special attribute `spec.restartedAt` that you need to update in order to trigger a new Run execution. For example:

Reviewer: ```suggestion
Yes. There is a special attribute `spec.restartedAt` that you need to update in order to trigger a new Run execution. For example:
```

	```console
  $ kubectl patch module <NAME> --type=merge --patch '{"spec": {"restartedAt": "'`date -u -Iseconds`'"}}'
  ```

## Agent Pool Controller

- **Where can I find Agent tokens?**

>   The Agent tokens are sensitive and will be saved in Kubernetes Secret. The name of the Kubernetes Secret object will be generated automatically and has the following pattern: `<CR.metadata.name>-agent-pool`.

Reviewer: ```suggestion
  The Agent tokens are sensitive and will be saved in a Secret. The name of the Secret object will be generated automatically and has the following pattern: `<metadata.name>-agent-pool`.
```

> - **Does the Operator restore tokens, if I delete the whole Kubernetes Secret with Agent Tokens or a single token from it?**

Reviewer: ```suggestion
- **Does the Operator restore tokens if I delete the whole Secret containing the Agent Tokens or a single token from it?**
```
